File: lib/apps/vocabulary/services/vocabulary_audio_service.dart
import '../../../services/audio_generation_service.dart';
import '../../../services/logger.dart';
import '../../../services/queue_consumer.dart';
import '../../../services/queue_message.dart';
import '../../../apps/file_system/models/file_item.dart';
import 'vocabulary_storage.dart';

/// Service that listens to vocabulary-audio queue and generates audio for words and samples.
/// Extends QueueConsumer to get both push notifications and polling for reliability.
class VocabularyAudioService extends QueueConsumer {
  static VocabularyAudioService? _instance;
  static VocabularyAudioService get instance =>
      _instance ??= VocabularyAudioService._();

  VocabularyAudioService._();

  final Logger _logger = Logger(
    appId: 'vocabulary',
    appName: 'Vocabulary Audio Service',
  );

  @override
  String get id => 'vocabulary-audio-consumer';

  @override
  String get name => 'Vocabulary Audio Consumer';

  @override
  String get queueId => 'vocabulary-audio';

  @override
  void onStart() {
    _logger.info(
      'Vocabulary audio service started',
      eventType: 'service_start',
    );
  }

  @override
  void onStop() {
    _logger.info(
      'Vocabulary audio service stopped',
      eventType: 'service_stop',
    );
  }

  @override
  void onError(QueueMessage message, Object error) {
    _logger.error(
      'Error processing message: $error',
      eventType: 'message_error',
      metadata: {'messageId': message.id},
    );
  }

  /// Process a message from the queue
  @override
  Future<bool> processMessage(QueueMessage message) async {
    try {
      final wordId = message.payload['wordId'] as String?;

      await _logger.info(
        'Processing vocabulary audio message',
        eventType: 'message_receive',
        metadata: {
          'messageId': message.id,
          'wordId': wordId,
          'eventType': message.eventType,
          'deliveryCount': message.deliveryCount,
        },
      );

      if (wordId == null || wordId.isEmpty) {
        await _logger.warning(
          'Invalid message payload - missing wordId',
          eventType: 'message_invalid',
          metadata: {'wordId': wordId},
        );
        return true; // Acknowledge to remove from queue
      }

      // Retrieve word from database
      final word = await VocabularyStorage.instance.getWord(wordId);
      if (word == null) {
        await _logger.warning(
          'Word no longer exists, skipping audio generation',
          eventType: 'word_not_found',
          metadata: {'wordId': wordId},
        );
        return true;
      }

      // Verify word has meaning and samples
      if (word.meaning.isEmpty || word.samplePhrases.isEmpty) {
        await _logger.warning(
          'Word has no meaning or samples, skipping audio generation',
          eventType: 'word_incomplete',
          metadata: {
            'wordId': wordId,
            'word': word.word,
            'hasMeaning': word.meaning.isNotEmpty,
            'sampleCount': word.samplePhrases.length,
          },
        );
        return true; // Acknowledge to remove from queue
      }

      final wordText = word.word;
      final samplePhrases = word.samplePhrases;

      final audioService = AudioGenerationService.instance;

      if (!audioService.isConfigured) {
        await _logger.error(
          'Audio generation service not configured',
          eventType: 'audio_not_configured',
          metadata: {
            'suggestion': 'Set the audio generation API URL in Settings',
          },
        );
        return false; // Will retry
      }

      // Generate audio for the word
      String? wordAudioPath;
      try {
        await _logger.info(
          'Generating audio for word',
          eventType: 'word_audio_start',
          metadata: {'wordId': wordId, 'word': wordText},
        );

        final wordAudioFile = await audioService.generateAudioForApp(
          text: wordText,
          filename: '${wordText}_word',
          appId: 'vocabulary',
        );
        wordAudioPath = wordAudioFile.relativePath;

        await _logger.info(
          'Successfully generated word audio',
          eventType: 'word_audio_success',
          metadata: {
            'wordId': wordId,
            'word': wordText,
            'audioPath': wordAudioPath,
          },
        );
      } catch (e) {
        await _logger.error(
          'Failed to generate word audio: $e',
          eventType: 'word_audio_failed',
          metadata: {'wordId': wordId, 'word': wordText, 'error': e.toString()},
        );
        // Continue with sample phrases even if word audio fails
      }

      // Generate audio for sample phrases
      final phraseAudioPaths = <String>[];
      if (samplePhrases != null && samplePhrases.isNotEmpty) {
        for (var i = 0; i < samplePhrases.length; i++) {
          final phrase = samplePhrases[i];
          if (phrase.isEmpty) continue;

          try {
            await _logger.info(
              'Generating audio for sample phrase',
              eventType: 'phrase_audio_start',
              metadata: {
                'wordId': wordId,
                'phraseIndex': i,
                'phrasePreview': phrase.length > 50
                    ? '${phrase.substring(0, 50)}...'
                    : phrase,
              },
            );

            final phraseAudioFile = await audioService.generateAudioForApp(
              text: phrase,
              filename: '${wordText}_phrase_${i + 1}',
              appId: 'vocabulary',
            );
            phraseAudioPaths.add(phraseAudioFile.relativePath);

            await _logger.info(
              'Successfully generated phrase audio',
              eventType: 'phrase_audio_success',
              metadata: {
                'wordId': wordId,
                'phraseIndex': i,
                'audioPath': phraseAudioFile.relativePath,
              },
            );
          } catch (e) {
            await _logger.error(
              'Failed to generate phrase audio: $e',
              eventType: 'phrase_audio_failed',
              metadata: {
                'wordId': wordId,
                'phraseIndex': i,
                'error': e.toString(),
              },
            );
            // Add empty string to maintain index alignment
            phraseAudioPaths.add('');
          }
        }
      }

      // Update word with audio paths
      await VocabularyStorage.instance.updateWordAudio(
        wordId,
        wordAudioPath: wordAudioPath,
        sampleAudioPaths: phraseAudioPaths,
      );

      await _logger.info(
        'Successfully updated word with audio paths',
        eventType: 'audio_save',
        metadata: {
          'wordId': wordId,
          'word': wordText,
          'hasWordAudio': wordAudioPath != null,
          'phrasesCount': phraseAudioPaths.length,
        },
      );

      return true;
    } catch (e, stackTrace) {
      await _logger.error(
        'Error processing vocabulary audio message: $e',
        eventType: 'message_error',
        metadata: {
          'messageId': message.id,
          'error': e.toString(),
          'stackTrace': stackTrace.toString().split('\n').take(5).join('\n'),
        },
      );
      return false; // Will retry
    }
  }

  /// Dispose the service
  void dispose() {
    stop();
    _logger.info(
      'Vocabulary audio service disposed',
      eventType: 'service_dispose',
    );
  }

  /// Reset instance for testing
  static void resetInstance() {
    _instance?.dispose();
    _instance = null;
  }
}

File: playground/Playground/Playground/Sources/Services/VocabularyAudioService.swift
import Foundation

/// Service for generating audio for vocabulary words and sample phrases
/// Matches the Dart VocabularyAudioService implementation
class VocabularyAudioService {
    static let shared = VocabularyAudioService()
    private let audioService = AudioGenerationService.shared
    private let storage = VocabularyStorage.shared

    private init() {}

    // MARK: - Audio Generation

    /// Generate audio for a vocabulary word and its sample phrases
    /// - Parameter wordId: ID of the vocabulary word
    /// - Returns: True if successful, false if failed
    @discardableResult
    func generateAudioForWord(wordId: String) async throws -> Bool {
        print("🎵 Generating audio for word: \(wordId)")

        // Retrieve word from database
        guard let word = try await storage.getWord(id: wordId).get() else {
            print("⚠️ Word not found: \(wordId)")
            return false
        }

        // Verify word has meaning and samples
        if word.meaning.isEmpty || word.samplePhrases.isEmpty {
            print("⚠️ Word has no meaning or samples, skipping audio generation")
            print("   Word: \(word.word)")
            print("   Has meaning: \(!word.meaning.isEmpty)")
            print("   Sample count: \(word.samplePhrases.count)")
            return false
        }

        let wordText = word.word
        let samplePhrases = word.samplePhrases

        // Check if service is configured
        guard audioService.isConfigured else {
            print("❌ Audio generation service not configured")
            print("   Suggestion: Set the audio generation API URL in Settings")
            throw AudioGenerationError.notConfigured
        }

        var wordAudioPath: String?
        var sampleAudioPaths: [String] = []

        // Generate audio for the word
        do {
            print("🎵 Generating audio for word: \(wordText)")

            let wordAudioURL = try await audioService.generateAudio(
                text: wordText,
                filename: "\(wordText)_word"
            )
            wordAudioPath = audioService.getRelativePath(for: wordAudioURL)

            print("✅ Successfully generated word audio: \(wordAudioPath ?? "")")
        } catch {
            print("❌ Failed to generate word audio: \(error)")
            // Continue with sample phrases even if word audio fails
        }

        // Generate audio for sample phrases
        for (index, phrase) in samplePhrases.enumerated() {
            guard !phrase.isEmpty else {
                sampleAudioPaths.append("")
                continue
            }

            do {
                print("🎵 Generating audio for sample phrase \(index + 1)/\(samplePhrases.count)")
                print("   Phrase: \(phrase.prefix(50))\(phrase.count > 50 ? "..." : "")")

                let phraseAudioURL = try await audioService.generateAudio(
                    text: phrase,
                    filename: "\(wordText)_phrase_\(index + 1)"
                )
                let phrasePath = audioService.getRelativePath(for: phraseAudioURL)
                sampleAudioPaths.append(phrasePath)

                print("✅ Successfully generated phrase audio: \(phrasePath)")
            } catch {
                print("❌ Failed to generate phrase audio \(index + 1): \(error)")
                // Add empty string to maintain index alignment
                sampleAudioPaths.append("")
            }
        }

        // Update word with audio paths
        try storage.updateWord(
            id: wordId,
            wordAudioPath: wordAudioPath,
            sampleAudioPaths: sampleAudioPaths
        ).get()

        print("✅ Successfully updated word with audio paths")
        print("   Word audio: \(wordAudioPath ?? "none")")
        print("   Phrase audios: \(sampleAudioPaths.count)")

        return true
    }

    /// Generate audio for multiple words in sequence
    /// - Parameter wordIds: Array of word IDs
    /// - Returns: Number of words successfully processed
    func generateAudioForWords(wordIds: [String]) async -> Int {
        var successCount = 0

        for wordId in wordIds {
            do {
                if try await generateAudioForWord(wordId: wordId) {
                    successCount += 1
                }
            } catch {
                print("❌ Error generating audio for word \(wordId): \(error)")
            }
        }

        return successCount
    }

    // MARK: - Audio Playback

    /// Get audio URL for a vocabulary word
    /// - Parameter word: The vocabulary word
    /// - Returns: URL of the word audio file, or nil if not available
    func getWordAudioURL(for word: VocabularyWord) -> URL? {
        guard let audioPath = word.wordAudioPath else {
            return nil
        }
        return audioService.getAbsoluteURL(from: audioPath)
    }

    /// Get audio URLs for sample phrases
    /// - Parameter word: The vocabulary word
    /// - Returns: Array of URLs (may contain nils for missing audio)
    func getSampleAudioURLs(for word: VocabularyWord) -> [URL?] {
        return word.sampleAudioPaths.map { path in
            guard !path.isEmpty else { return nil }
            return audioService.getAbsoluteURL(from: path)
        }
    }

    // MARK: - Utility

    /// Check if a word has audio
    /// - Parameter word: The vocabulary word
    /// - Returns: True if word has at least word audio or one sample audio
    func hasAudio(for word: VocabularyWord) -> Bool {
        if word.wordAudioPath != nil && !word.wordAudioPath!.isEmpty {
            return true
        }

        return word.sampleAudioPaths.contains { !$0.isEmpty }
    }

    /// Delete audio files for a word
    /// - Parameter word: The vocabulary word
    func deleteAudio(for word: VocabularyWord) {
        // Delete word audio
        if let wordPath = word.wordAudioPath, !wordPath.isEmpty {
            if let url = audioService.getAbsoluteURL(from: wordPath) {
                try? FileManager.default.removeItem(at: url)
                print("🗑️ Deleted word audio: \(wordPath)")
            }
        }

        // Delete sample audios
        for path in word.sampleAudioPaths where !path.isEmpty {
            if let url = audioService.getAbsoluteURL(from: path) {
                try? FileManager.default.removeItem(at: url)
                print("🗑️ Deleted sample audio: \(path)")
            }
        }
    }
}
